<?xml version="1.0"?>
<robot name="Opanoid_sketch">

<!--materials-->
  <material name="blue">
    <color rgba="0 0 0.8 1"/>
  </material>

  <material name="white">
    <color rgba="1 1 1 1"/>
  </material>
  
<!--links-->
  <link name="base_link">
  	<visual>
  		<geometry>
				<box size="0.15 0.35 0.15"/>
			</geometry>
		</visual>
		<collision>
  		<geometry>
				<box size="0.15 0.35 0.15"/>
			</geometry>
		</collision>
		<inertial>
      <mass value="10"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
	</link>
	<link name="l_pelvis_bracket_2">
  	<visual>
  		<geometry>
				<box size="0.01 0.01 0.01"/>
			</geometry>
			<origin xyz="0 0 0" rpy="0 0 0"/>
		</visual>
		<collision>
  		<geometry>
				<box size="0.01 0.01 0.01"/>
			</geometry>
			<origin xyz="0 0 0" rpy="0 0 0"/>
		</collision>
		<inertial>
      <mass value="10"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>
	<link name="l_pelvis_attachment">
  	<visual>
  		<geometry>
				<box size="0.1 0.1 0.4"/>
			</geometry>
			<origin xyz="0 0 -0.2" rpy="0 0 0"/>
			<material name="white"/>
		</visual>
		<collision>
  		<geometry>
				<box size="0.1 0.1 0.4"/>
			</geometry>
			<origin xyz="0 0 -0.2" rpy="0 0 0"/>
		</collision>
		<inertial>
      <mass value="10"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>
	<link name="l_ankle_2">
  	<visual>
  		<geometry>
				<box size="0.07 0.07 0.5"/>
			</geometry>
			<origin xyz="0 0 -0.25" rpy="0 0 0"/>
			<material name="blue"/>
		</visual>
		<collision>
  		<geometry>
				<box size="0.07 0.07 0.5"/>
			</geometry>
			<origin xyz="0 0 -0.25" rpy="0 0 0"/>
		</collision>
		<inertial>
      <mass value="10"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>
  <link name="l_ankle_1">
  	<visual>
  		<geometry>
				<box size="0.01 0.01 0.01"/>
			</geometry>
			<origin xyz="0 0 0" rpy="0 0 0"/>
		</visual>
		<collision>
  		<geometry>
				<box size="0.01 0.01 0.01"/>
			</geometry>
			<origin xyz="0 0 0" rpy="0 0 0"/>
		</collision>
		<inertial>
      <mass value="10"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>
  <link name="leftFoot">
  	<visual>
  		<geometry>
				<box size="0.175 0.07 0.03"/>
			</geometry>
			<origin xyz="0.0525 0 0" rpy="0 0 0"/>
			<material name="white"/>
		</visual>
		<collision>
  		<geometry>
				<box size="0.175 0.07 0.03"/>
			</geometry>
			<origin xyz="0.0525 0 0" rpy="0 0 0"/>
		</collision>
		<inertial>
      <mass value="10"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>
  <link name="leftToe">
  	<visual>
  		<geometry>
				<box size="0.03 0.07 0.03"/>
			</geometry>
			<origin xyz="0.015 0 0" rpy="0 0 0"/>
			<material name="blue"/>
		</visual>
		<collision>
  		<geometry>
				<box size="0.03 0.07 0.03"/>
			</geometry>
			<origin xyz="0.015 0 0" rpy="0 0 0"/>
		</collision>
		<inertial>
      <mass value="10"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>
  <link name="r_pelvis_bracket_2">
  	<visual>
  		<geometry>
				<box size="0.01 0.01 0.01"/>
			</geometry>
			<origin xyz="0 0 0" rpy="0 0 0"/>
		</visual>
		<collision>
  		<geometry>
				<box size="0.01 0.01 0.01"/>
			</geometry>
			<origin xyz="0 0 0" rpy="0 0 0"/>
		</collision>
		<inertial>
      <mass value="10"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>
  <link name="r_pelvis_attachment">
  	<visual>
  		<geometry>
				<box size="0.1 0.1 0.4"/>
			</geometry>
			<origin xyz="0 0 -0.2" rpy="0 0 0"/>
			<material name="white"/>
		</visual>
		<collision>
  		<geometry>
				<box size="0.1 0.1 0.4"/>
			</geometry>
			<origin xyz="0 0 -0.2" rpy="0 0 0"/>
		</collision>
		<inertial>
      <mass value="10"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>
  <link name="r_ankle_2">
  	<visual>
  		<geometry>
				<box size="0.07 0.07 0.5"/>
			</geometry>
			<origin xyz="0 0 -0.25" rpy="0 0 0"/>
			<material name="blue"/>
		</visual>
		<collision>
  		<geometry>
				<box size="0.07 0.07 0.5"/>
			</geometry>
			<origin xyz="0 0 -0.25" rpy="0 0 0"/>
		</collision>
		<inertial>
      <mass value="10"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>
  <link name="r_ankle_1">
  	<visual>
  		<geometry>
				<box size="0.01 0.01 0.01"/>
			</geometry>
			<origin xyz="0 0 0" rpy="0 0 0"/>
		</visual>
		<collision>
  		<geometry>
				<box size="0.01 0.01 0.01"/>
			</geometry>
			<origin xyz="0 0 0" rpy="0 0 0"/>
		</collision>
		<inertial>
      <mass value="10"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>
  <link name="rightFoot">
  	<visual>
  		<geometry>
				<box size="0.175 0.07 0.03"/>
			</geometry>
			<origin xyz="0.0525 0 0" rpy="0 0 0"/>
			<material name="white"/>
		</visual>
		<collision>
  		<geometry>
				<box size="0.175 0.07 0.03"/>
			</geometry>
			<origin xyz="0.0525 0 0" rpy="0 0 0"/>
		</collision>
		<inertial>
      <mass value="10"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>
  <link name="rightToe">
  	<visual>
  		<geometry>
				<box size="0.03 0.07 0.03"/>
			</geometry>
			<origin xyz="0.015 0 0" rpy="0 0 0"/>
			<material name="blue"/>
		</visual>
		<collision>
  		<geometry>
				<box size="0.03 0.07 0.03"/>
			</geometry>
			<origin xyz="0.015 0 0" rpy="0 0 0"/>
		</collision>
		<inertial>
      <mass value="10"/>
      <inertia ixx="0.4" ixy="0.0" ixz="0.0" iyy="0.4" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>
  <link name="imu_link">
  	<collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
    		<box size="0 0 0"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
    		<box size="0.07 0.07 0.07"/>
      </geometry>
      <material name="white"/>
    </visual>
    <inertial>
      <mass value="1e-5" />
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <inertia ixx="1e-6" ixy="0" ixz="0" iyy="1e-6" iyz="0" izz="1e-6" />
    </inertial>
  </link>

<!--joints-->
	<joint name="leftThighY" type="revolute">
    <parent link="base_link"/>
    <child link="l_pelvis_bracket_2"/>
    <axis xyz="0 -1 0"/>
    <limit effort="100.0" lower="-1.570" upper="1.570" velocity="7.48"/>
		<origin xyz="0 0.12 -0.075" rpy="0 0 0"/>
  </joint>
  <transmission name="leftThighY_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$leftThighY_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="leftThighY">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>
  
  <joint name="leftThighX" type="revolute">
    <parent link="l_pelvis_bracket_2"/>
    <child link="l_pelvis_attachment"/>
    <axis xyz="1 0 0"/>
    <limit effort="100.0" lower="-0.15" upper="1.570" velocity="6.546"/>
		<origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>
  <transmission name="leftThighX_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$leftThighX_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="leftThighX">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>
  
	<joint name="leftKnee" type="revolute">
    <parent link="pelvis_attachment"/>
    <child link="l_ankle_2"/>
    <axis xyz="0 1 0"/>
    <limit effort="100.0" lower="0" upper="1.570" velocity="7.48"/>
		<origin xyz="0 0 -0.4" rpy="0 0 0"/>
  </joint>
  <transmission name="leftKnee_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$leftKnee_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="leftKnee">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>
  
  <joint name="leftAnkleX" type="revolute">
    <parent link="l_ankle_2"/>
    <child link="l_ankle_1"/>
    <axis xyz="1 0 0"/>
    <limit effort="100.0" lower="-1.570" upper="1.570" velocity="7.48"/>
		<origin xyz="0 0 -0.5" rpy="0 0 0"/>
  </joint>
  <transmission name="leftAnkleX_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$leftAnkleX_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="leftAnkleX">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>
  
  <joint name="leftAnkleYGODDAMNIT" type="revolute">
    <parent link="l_ankle_1"/>
    <child link="leftFoot"/>
    <axis xyz="0 1 0"/>
    <limit effort="100.0" lower="-1.570" upper="1.570" velocity="7.48"/>
		<origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>
  <transmission name="leftAnkleY_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$leftAnkleY_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="leftAnkleY">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>
   
  <joint name="leftToe" type="revolute">
    <parent link="leftFoot"/>
    <child link="leftToe"/>
    <axis xyz="0 1 0"/>
    <limit effort="100.0" lower="-1.570" upper="1.570" velocity="7.48"/>
		<origin xyz="0.14 0 0" rpy="0 0 0"/>
  </joint>
  <transmission name="leftToe_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$leftToe_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="leftToe">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>
  
  <joint name="rightThighY" type="revolute">
    <parent link="base_link"/>
    <child link="r_pelvis_bracket_2"/>
    <axis xyz="0 -1 0"/>
    <limit effort="1000.0" lower="-1.570" upper="1.570" velocity="7.48"/>
		<origin xyz="0 -0.12 -0.075" rpy="0 0 0"/>
  </joint>
  <transmission name="rightThighY_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$rightThighY_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="rightThighY">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>
  
  <joint name="rightThighX" type="revolute">
    <parent link="r_pelvis_bracket_2"/>
    <child link="r_pelvis_attachment"/>
    <axis xyz="-1 0 0"/>
    <limit effort="100.0" lower="-0.15" upper="1.5705" velocity="6.546"/>
		<origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>
  <transmission name="rightThighX_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$rightThighX_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="rightThighX">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>

	<joint name="rightKnee" type="revolute">
    <parent link="r_pelvis_attachment"/>
    <child link="r_ankle_2"/>
    <axis xyz="0 1 0"/>
    <limit effort="100.0" lower="0" upper="1.570" velocity="7.48"/>
		<origin xyz="0 0 -0.4" rpy="0 0 0" />
  </joint>
  <transmission name="rightKnee_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$rightKnee_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="rightKnee">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>
  
  <joint name="rightAnkleX" type="revolute">
    <parent link="r_ankle_2"/>
    <child link="r_ankle_1"/>
    <axis xyz="1 0 0"/>
    <limit effort="100.0" lower="-1.570" upper="1.570" velocity="7.48"/>
		<origin xyz="0 0 -0.5" rpy="0 0 0"/>
  </joint>
  <transmission name="rightAnkleX_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$rightAnkleX_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="rightAnkleX">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>
  
  <joint name="rightAnkleY" type="revolute">
    <parent link="r_ankle_1"/>
    <child link="rightFoot"/>
    <axis xyz="0 1 0"/>
    <limit effort="100.0" lower="-1.570" upper="1.570" velocity="7.48"/>
		<origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>
  <transmission name="rightAnkleY_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$rightAnkleY_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="rightAnkleY">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>
  
  <joint name="rightToe" type="revolute">
    <parent link="rightFoot"/>
    <child link="rightToe"/>
    <axis xyz="0 1 0"/>
    <limit effort="100.0" lower="-1.570" upper="1.570" velocity="7.48"/>
		<origin xyz="0.14 0 0" rpy="0 0 0"/>
  </joint>
  <transmission name="rightToe_trans">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="$rightToe_motor">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="rightToe">
      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
    </joint>
  </transmission>
  
  <joint name="imu_joint" type="fixed">
    <axis xyz="0 1 0" />
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="imu_link"/>
  </joint>
  
  <gazebo reference="imu_link">
    <gravity>true</gravity>
    <sensor name="imu_sensor" type="imu">
      <always_on>true</always_on>
      <update_rate>100</update_rate>
      <visualize>true</visualize>
      <topic>__default_topic__</topic>
      <plugin filename="libgazebo_ros_imu_sensor.so" name="imu_plugin">
        <topicName>imu</topicName>
        <bodyName>imu_link</bodyName>
        <updateRateHZ>10.0</updateRateHZ>
        <gaussianNoise>0.0</gaussianNoise>
        <xyzOffset>0 0 0</xyzOffset>
        <rpyOffset>0 0 0</rpyOffset>
        <frameName>imu_link</frameName>
      </plugin>
      <pose>0 0 0 0 0 0</pose>
    </sensor>
  </gazebo>
  
  <gazebo>
    <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
      <robotNamespace>/</robotNamespace>
    </plugin>
  </gazebo> 
  
</robot>
